torch/distributed/_composable/offload.py

```diff
@@ -0,0 +1,124 @@
+from enum import auto, Enum
+from typing import List, Optional
+
+import torch
+
+import torch.nn as nn
+
+from torch.distributed._composable_state import _State
+from .contract import contract
+
+
+class _DeviceState(Enum):
+    CPU = auto()
+    GPU_UNSYNCED = auto()
+    GPU_SYNCED = auto()
+
+
+@contract()
+def frozen_offload(
+    module: nn.Module,
+):
+    state = frozen_offload.state(module)
+    _check_module_params(module)
+    _init_data(state, module)
+    state._forward_handle = module.register_forward_hook(to_cpu)
+    state._forward_pre_handle = module.register_forward_pre_hook(_to_gpu_synced)
+
+
+def to_cpu(module: nn.Module, *unused_args, **unused_kwargs):
+    """
+    Frees the GPU parameter memory.
+    """
+    state = frozen_offload.state(module)
+    if state is None:
+        return
+    if state._device_state == _DeviceState.CPU:
+        return
+    with torch.profiler.record_function("frozen_offload::to_cpu"):
+        # Choose to not persist any changes from the GPU copy to the CPU copy
+        for param, cpu_view in zip(state._params, state._cpu_views):
+            param.data = cpu_view
+        # state._gpu_flat_tensor.untyped_storage().resize_(0)
+        state._gpu_flat_tensor.record_stream(torch.cuda.current_stream())
+        state._gpu_flat_tensor = None
+        state._device_state = _DeviceState.CPU
+
+
+def to_gpu(module: nn.Module, *unused_args, **unused_kwargs):
+    """
+    Allocates GPU memory for the parameters and copies them from CPU to GPU in
+    a separate CUDA stream. This must be followed with a ``wait_on_copy()``
```

Comment:
what is wait_on_copy? Don't see an example in the test

Reply:
My bad. I refactored the approach so that the `wait_stream()` is in `_to_gpu_synced()` instead of being a separate function.